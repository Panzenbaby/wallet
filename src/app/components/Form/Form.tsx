import React from "react";
import { FormProvider, SubmitHandler, UseFormMethods } from "react-hook-form";

type FormProperties<T = any> = {
	onSubmit: SubmitHandler<T>;
	context: UseFormMethods<T>;
} & Omit<React.FormHTMLAttributes<any>, "onSubmit">;

export const Form = React.forwardRef<HTMLFormElement, FormProperties>(
	({ children, context, onSubmit, ...properties }: FormProperties, reference) => (
		<FormProvider {...context}>
			<form
				data-testid="Form"
				ref={reference}
				className="space-y-5"
				onSubmit={context.handleSubmit(onSubmit)}
				{...properties}
			>
				{children}
			</form>
		</FormProvider>
	),
);

Form.displayName = "Form";
