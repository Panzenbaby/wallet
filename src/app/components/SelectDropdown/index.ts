export * from "./SelectDropdown";
export * from "./SelectDropdown.models";
