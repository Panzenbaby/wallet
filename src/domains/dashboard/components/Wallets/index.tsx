export * from "./hooks";
export * from "./Wallets";
export * from "./Wallets.contracts";
export * from "./WalletsGrid";
export * from "./WalletsList";
