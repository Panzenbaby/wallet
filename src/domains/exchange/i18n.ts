export const translations: { [key: string]: any } = {
	EXCHANGE_NAME: "Exchange Name",

	MODAL_ADD_EXCHANGE: {
		TITLE: "Exchange",
	},

	PAGE_EXCHANGES: {
		ADD_EXCHANGE: "Add Exchange",
		EMPTY_MESSAGE: "There are currently no exchange providers available, please try again at a later time.",
		SUBTITLE: "Choose one of the exchange providers below to swap directly from within the wallet.",
		TITLE: "Exchange",
	},

	SELECT_EXCHANGE_MESSAGE: "Select one of the exchanges you have installed",

	YOUR_EXCHANGE_LIST: "Your Exchange List",
};
