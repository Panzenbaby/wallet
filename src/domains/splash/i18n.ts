export const translations: { [key: string]: any } = {
	BRAND: "Payvo Wallet",
	COPYRIGHT: "Â© ARK.io",
	LOADING: "Initializing...",
	PRODUCT: "An ARK.io Product",
	RIGHTS: "All Rights Reserved",
	VERSION: "Version",
};
