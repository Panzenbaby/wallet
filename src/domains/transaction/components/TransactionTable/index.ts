export * from "./TransactionTable";
export * from "./TransactionTable.contracts";
export * from "./TransactionTable.helpers";
export * from "./TransactionTable.models";
