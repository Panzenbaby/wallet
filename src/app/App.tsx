import "focus-visible";

// import { TRX } from "@payvo/sdk-trx";
// import { XLM } from "@payvo/sdk-xlm";
// import { XRP } from "@payvo/sdk-xrp";
// import { ZIL } from "@payvo/sdk-zil";
import LedgerTransportNodeHID from "@ledgerhq/hw-transport-node-hid-singleton";
// import { LUNA } from "@payvo/sdk-luna";
// import { NANO } from "@payvo/sdk-nano";
// import { NEO } from "@payvo/sdk-neo";
import { Environment } from "@payvo/profiles";
// import { ADA } from "@payvo/sdk-ada";
import { ARK } from "@payvo/sdk-ark";
// import { ATOM } from "@payvo/sdk-atom";
// import { AVAX } from "@payvo/sdk-avax";
// import { BTC } from "@payvo/sdk-btc";
// import { DOT } from "@payvo/sdk-dot";
// import { EGLD } from "@payvo/sdk-egld";
// import { ETH } from "@payvo/sdk-eth";
import { LSK } from "@payvo/sdk-lsk";
import { Offline } from "domains/error/pages";
import { Splash } from "domains/splash/pages";
import { usePluginManagerContext } from "plugins/context/PluginManagerProvider";
import React, { useLayoutEffect, useState } from "react";
import { useErrorHandler } from "react-error-boundary";
import { I18nextProvider } from "react-i18next";
import { useHistory } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import { StubStorage } from "tests/mocks";
import { setThemeSource, shouldUseDarkColors } from "utils/electron-utils";
import { bootEnvWithProfileFixtures, isE2E, isUnit } from "utils/test-helpers";

import { middlewares, RouterView, routes } from "../router";
import { ConfigurationProvider, EnvironmentProvider, LedgerProvider, useEnvironmentContext } from "./contexts";
import { useDeeplink, useNetworkStatus, useProfileSynchronizer } from "./hooks";
import { i18n as index18n } from "./i18n";
import { PluginProviders } from "./PluginProviders";
import { SentryProvider } from "./sentry/SentryProvider";
import { SentryRouterWrapper } from "./sentry/SentryRouterWrapper";
import { httpClient } from "./services";

const RouteWrappers = ({ children }: { children: React.ReactNode }) => (
	<SentryRouterWrapper>{children}</SentryRouterWrapper>
);

const Main = () => {
	const [showSplash, setShowSplash] = useState(true);
	const { env } = useEnvironmentContext();
	const { loadPlugins } = usePluginManagerContext();
	const isOnline = useNetworkStatus();
	const history = useHistory();

	useProfileSynchronizer({
		onProfileRestoreError: () => history.push("/"),
	});

	useDeeplink();

	useLayoutEffect(() => {
		setThemeSource("system");

		document.body.classList.remove(`theme-${shouldUseDarkColors() ? "light" : "dark"}`);
		document.body.classList.add(`theme-${shouldUseDarkColors() ? "dark" : "light"}`);
	}, []);

	const handleError = useErrorHandler();

	useLayoutEffect(() => {
		const boot = async () => {
			try {
				/* istanbul ignore next */
				if (isE2E() || isUnit()) {
					await bootEnvWithProfileFixtures({ env, shouldRestoreDefaultProfile: isUnit() });

					loadPlugins();
					setShowSplash(false);
					return;
				}

				/* istanbul ignore next */
				await env.verify();
				/* istanbul ignore next */
				await env.boot();
				/* istanbul ignore next */
				await loadPlugins();
			} catch (error) {
				handleError(error);
			}

			setShowSplash(false);
		};

		boot();
	}, [env, handleError, loadPlugins]);

	const renderContent = () => {
		if (showSplash) {
			return <Splash />;
		}

		if (!isOnline) {
			return <Offline />;
		}

		return <RouterView routes={routes} middlewares={middlewares} wrapper={RouteWrappers} />;
	};

	return (
		<main data-testid="Main">
			<ToastContainer closeOnClick={false} newestOnTop />

			{renderContent()}
		</main>
	);
};

export const App = () => {
	/**
	 * Ensure that the Environment object will not be recreated when the state changes,
	 * as the data is stored in memory by the `DataRepository`.
	 */

	/* istanbul ignore next */
	const storage = isE2E() || isUnit() ? new StubStorage() : "indexeddb";

	const [environment] = useState(
		() =>
			new Environment({
				coins: {
					// ADA,
					ARK,
					// ATOM,
					// AVAX,
					// BTC,
					// DOT,
					// ETH,
					// EGLD,
					LSK,
					// NEO,
					// NANO,
					// LUNA,
					// TRX,
					// XLM,
					// XRP,
					// ZIL,
				},
				httpClient,
				storage,
			}),
	);

	return (
		<I18nextProvider i18n={index18n}>
			<EnvironmentProvider env={environment}>
				<ConfigurationProvider defaultConfiguration={{ profileIsSyncingExchangeRates: true }}>
					<SentryProvider>
						<LedgerProvider transport={LedgerTransportNodeHID}>
							<PluginProviders>
								<Main />
							</PluginProviders>
						</LedgerProvider>
					</SentryProvider>
				</ConfigurationProvider>
			</EnvironmentProvider>
		</I18nextProvider>
	);
};
