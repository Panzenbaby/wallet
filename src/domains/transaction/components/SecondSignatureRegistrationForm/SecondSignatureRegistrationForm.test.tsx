/* eslint-disable @typescript-eslint/require-await */
import { BIP39 } from "@payvo/cryptography";
import { Contracts as ProfilesContracts } from "@payvo/profiles";
import { Contracts } from "@payvo/sdk";
import { RenderResult, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Form } from "app/components/Form";
import { toasts } from "app/services";
import React, { useEffect } from "react";
import { useForm, UseFormMethods } from "react-hook-form";
import { useTranslation } from "react-i18next";
import secondSignatureFixture from "tests/fixtures/coins/ark/devnet/transactions/second-signature-registration.json";
import * as utils from "utils/electron-utils";
import { env, fireEvent, getDefaultProfileId, MNEMONICS, render, screen, waitFor } from "utils/testing-library";

import { translations as transactionTranslations } from "../../i18n";
import { SecondSignatureRegistrationForm, signSecondSignatureRegistration } from "./SecondSignatureRegistrationForm";

describe("SecondSignatureRegistrationForm", () => {
	const passphrase = "power return attend drink piece found tragic fire liar page disease combine";
	let profile: ProfilesContracts.IProfile;
	let wallet: ProfilesContracts.IReadWriteWallet;

	const fees = { avg: 1.354, max: 10, min: 0, static: 0 };

	beforeEach(() => {
		profile = env.profiles().findById(getDefaultProfileId());
		wallet = profile.wallets().first();
	});

	const createTransactionMock = (wallet: ProfilesContracts.IReadWriteWallet) =>
		// @ts-ignore
		jest.spyOn(wallet.transaction(), "transaction").mockReturnValue({
			amount: () => secondSignatureFixture.data.amount / 1e8,
			data: () => ({ data: () => secondSignatureFixture.data }),
			explorerLink: () => `https://dexplorer.ark.io/transaction/${secondSignatureFixture.data.id}`,
			fee: () => secondSignatureFixture.data.fee / 1e8,
			id: () => secondSignatureFixture.data.id,
			recipient: () => secondSignatureFixture.data.recipient,
			sender: () => secondSignatureFixture.data.sender,
		});

	const renderComponent = (properties?: any) => {
		let form: UseFormMethods | undefined;

		const defaultValues = properties?.defaultValues ?? { fees };
		const onSubmit = properties?.onSubmit ?? jest.fn();
		const activeTab = properties?.activeTab ?? 1;

		const Component = () => {
			form = useForm<any>({ defaultValues, mode: "onChange" });

			const { register } = form;

			useEffect(() => {
				register("fee");
				register("fees");
				register("inputFeeSettings");
			}, [register]);

			return (
				<Form context={form} onSubmit={onSubmit}>
					<SecondSignatureRegistrationForm.component
						profile={profile}
						activeTab={activeTab}
						wallet={wallet}
					/>
				</Form>
			);
		};

		const renderResult: RenderResult = render(<Component />);
		return { ...renderResult, form };
	};

	it("should render generation step", async () => {
		const passphrase = "mock bip39 passphrase";
		const bip39GenerateMock = jest.spyOn(BIP39, "generate").mockReturnValue(passphrase);

		const { form, asFragment } = renderComponent();

		expect(asFragment()).toMatchSnapshot();

		await waitFor(() => expect(screen.getByTestId("SecondSignatureRegistrationForm__generation-step")));
		await waitFor(() => expect(form?.getValues("secondMnemonic")).toEqual(passphrase));

		expect(bip39GenerateMock).toHaveBeenCalled();

		bip39GenerateMock.mockRestore();
	});

	it("should not generate mnemonic if already set", async () => {
		const bip39GenerateMock = jest.spyOn(BIP39, "generate").mockImplementation();

		const { form, asFragment } = renderComponent({
			defaultValues: {
				fees,
				secondMnemonic: "test mnemonic",
			},
		});

		expect(asFragment()).toMatchSnapshot();

		await waitFor(() => expect(screen.getByTestId("SecondSignatureRegistrationForm__generation-step")));

		expect(form?.getValues("secondMnemonic")).toEqual("test mnemonic");

		expect(bip39GenerateMock).not.toHaveBeenCalled();

		bip39GenerateMock.mockRestore();
	});

	it("should set fee", async () => {
		renderComponent();

		// simple

		await waitFor(() => expect(screen.getAllByRole("radio")[1]).toBeChecked());

		userEvent.click(within(screen.getByTestId("InputFee")).getAllByRole("radio")[2]);

		await waitFor(() => expect(screen.getAllByRole("radio")[2]).toBeChecked());

		// advanced

		userEvent.click(screen.getByText(transactionTranslations.INPUT_FEE_VIEW_TYPE.ADVANCED));

		await waitFor(() => expect(screen.getByTestId("InputCurrency")).toBeVisible());

		fireEvent.change(screen.getByTestId("InputCurrency"), {
			target: {
				value: "9",
			},
		});

		await waitFor(() => expect(screen.getByTestId("InputCurrency")).toHaveValue("9"));
	});

	describe("backup step", () => {
		it("should render", async () => {
			const { asFragment } = renderComponent({
				activeTab: 2,
				defaultValues: {
					fees,
					secondMnemonic: "test mnemonic",
					wallet: {
						address: () => "address",
					},
				},
			});

			await waitFor(() =>
				expect(screen.getByTestId("SecondSignatureRegistrationForm__backup-step")).toBeTruthy(),
			);

			const writeTextMock = jest.fn();
			const clipboardOriginal = navigator.clipboard;
			(navigator as any).clipboard = { writeText: writeTextMock };

			fireEvent.click(screen.getByTestId("SecondSignature__copy"));

			await waitFor(() => expect(writeTextMock).toHaveBeenCalledWith("test mnemonic"));

			(navigator as any).clipboard = clipboardOriginal;

			expect(asFragment()).toMatchSnapshot();
		});

		it("should show success toast on successful download", async () => {
			renderComponent({
				activeTab: 2,
				defaultValues: {
					fees,
					secondMnemonic: "test mnemonic",
					wallet: {
						address: () => "address",
					},
				},
			});

			jest.spyOn(utils, "saveFile").mockResolvedValueOnce("filePath");

			const toastSpy = jest.spyOn(toasts, "success");

			fireEvent.click(screen.getByTestId("SecondSignature__download"));

			await waitFor(() => expect(toastSpy).toHaveBeenCalled());

			toastSpy.mockRestore();
		});

		it("should not show success toast on cancelled download", async () => {
			renderComponent({
				activeTab: 2,
				defaultValues: {
					fees,
					secondMnemonic: "test mnemonic",
					wallet: {
						address: () => "address",
					},
				},
			});

			jest.spyOn(utils, "saveFile").mockResolvedValueOnce(undefined);

			const toastSpy = jest.spyOn(toasts, "success");

			fireEvent.click(screen.getByTestId("SecondSignature__download"));

			await waitFor(() => expect(toastSpy).not.toHaveBeenCalled());

			toastSpy.mockRestore();
		});

		it("should show error toast on error", async () => {
			renderComponent({
				activeTab: 2,
				defaultValues: {
					fees,
					secondMnemonic: "test mnemonic",
					wallet: {
						address: () => "address",
					},
				},
			});

			jest.spyOn(utils, "saveFile").mockRejectedValueOnce(new Error("Error"));

			const toastSpy = jest.spyOn(toasts, "error");

			fireEvent.click(screen.getByTestId("SecondSignature__download"));

			await waitFor(() => expect(toastSpy).toHaveBeenCalled());

			toastSpy.mockRestore();
		});
	});

	it("should render verification step", async () => {
		const { form } = renderComponent({
			activeTab: 3,
			defaultValues: {
				fees,
				secondMnemonic: passphrase,
			},
		});

		await waitFor(() =>
			expect(screen.getByTestId("SecondSignatureRegistrationForm__verification-step")).toBeTruthy(),
		);

		expect(form?.getValues("verification")).toBeUndefined();

		const walletMnemonic = passphrase.split(" ");

		for (let index = 0; index < 3; index++) {
			const wordNumber = Number.parseInt(screen.getByText(/Select the/).innerHTML.replace(/Select the/, ""));

			fireEvent.click(screen.getByText(walletMnemonic[wordNumber - 1]));

			if (index < 2) {
				await waitFor(() => expect(screen.queryAllByText(/The (\d+)/).length === 2 - index));
			}
		}

		await waitFor(() => expect(form?.getValues("verification")).toBe(true));
	});

	it("should render review step", async () => {
		renderComponent({
			activeTab: 4,
			defaultValues: {
				fee: 0,
				fees,
			},
		});

		await waitFor(() => expect(screen.getByTestId("SecondSignatureRegistrationForm__review-step")).toBeTruthy());
	});

	it("should render transaction details", async () => {
		const DetailsComponent = () => {
			const { t } = useTranslation();
			return (
				<SecondSignatureRegistrationForm.transactionDetails
					translations={t}
					transaction={transaction}
					wallet={wallet}
				/>
			);
		};
		const transaction = {
			amount: () => secondSignatureFixture.data.amount / 1e8,
			data: () => ({ data: () => secondSignatureFixture.data }),
			fee: () => secondSignatureFixture.data.fee / 1e8,
			id: () => secondSignatureFixture.data.id,
			recipient: () => secondSignatureFixture.data.recipient,
			sender: () => secondSignatureFixture.data.sender,
		} as Contracts.SignedTransactionData;
		const { asFragment } = render(<DetailsComponent />);

		await waitFor(() => expect(screen.getByTestId("TransactionFee")).toBeTruthy());

		expect(asFragment()).toMatchSnapshot();
	});

	it("should sign transaction", async () => {
		const form = {
			clearErrors: jest.fn(),
			getValues: () => ({
				fee: "1",
				mnemonic: MNEMONICS[0],
				secondMnemonic: MNEMONICS[1],
				senderAddress: wallet.address(),
			}),
			setError: jest.fn(),
			setValue: jest.fn(),
		};
		const signMock = jest
			.spyOn(wallet.transaction(), "signSecondSignature")
			.mockReturnValue(Promise.resolve(secondSignatureFixture.data.id));
		const broadcastMock = jest.spyOn(wallet.transaction(), "broadcast").mockResolvedValue({
			accepted: [secondSignatureFixture.data.id],
			errors: {},
			rejected: [],
		});
		const transactionMock = createTransactionMock(wallet);

		await signSecondSignatureRegistration({
			env,
			form,
			profile,
		});

		expect(signMock).toHaveBeenCalled();
		expect(broadcastMock).toHaveBeenCalled();
		expect(transactionMock).toHaveBeenCalled();

		signMock.mockRestore();
		broadcastMock.mockRestore();
		transactionMock.mockRestore();
	});

	it("should sign transaction using encryption password", async () => {
		const walletUsesWIFMock = jest.spyOn(wallet.signingKey(), "exists").mockReturnValue(true);
		const walletWifMock = jest.spyOn(wallet.signingKey(), "get").mockReturnValue(passphrase);

		const form = {
			clearErrors: jest.fn(),
			getValues: () => ({
				encryptionPassword: "password",
				fee: "1",
				senderAddress: wallet.address(),
			}),
			setError: jest.fn(),
			setValue: jest.fn(),
		};
		const signMock = jest
			.spyOn(wallet.transaction(), "signSecondSignature")
			.mockReturnValue(Promise.resolve(secondSignatureFixture.data.id));
		const broadcastMock = jest.spyOn(wallet.transaction(), "broadcast").mockResolvedValue({
			accepted: [secondSignatureFixture.data.id],
			errors: {},
			rejected: [],
		});
		const transactionMock = createTransactionMock(wallet);

		await signSecondSignatureRegistration({
			env,
			form,
			profile,
		});

		expect(signMock).toHaveBeenCalled();
		expect(broadcastMock).toHaveBeenCalled();
		expect(transactionMock).toHaveBeenCalled();

		signMock.mockRestore();
		broadcastMock.mockRestore();
		transactionMock.mockRestore();
		walletUsesWIFMock.mockRestore();
		walletWifMock.mockRestore();
	});
});
