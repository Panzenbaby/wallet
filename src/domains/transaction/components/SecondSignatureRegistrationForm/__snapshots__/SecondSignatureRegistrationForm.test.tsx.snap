// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecondSignatureRegistrationForm backup step should render 1`] = `
<DocumentFragment>
  <form
    class="space-y-5"
    data-testid="Form"
  >
    <div>
      <div
        data-testid="tab-pabel__active-panel"
      >
        <section
          data-testid="SecondSignatureRegistrationForm__backup-step"
        >
          <div
            class="flex justify-between items-end bg-theme-background"
          >
            <div
              class="space-y-4"
            >
              <h1
                class="mb-0 text-2xl"
                data-testid="header__title"
              >
                Your Second Signature
              </h1>
            </div>
          </div>
          <div
            class="pt-6 space-y-6"
          >
            <div
              class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-warning-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center"
            >
              <div
                class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-warning-600"
              >
                <div
                  class="sc-bdfBwQ hjDozt"
                  height="20"
                  width="20"
                >
                  <svg>
                    alert-warning.svg
                  </svg>
                </div>
              </div>
              <div
                class="flex flex-col w-full"
              >
                <div
                  class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
                >
                  You are responsible for storing and protecting this mnemonic passphrase offline. The Payvo Wallet cannot reveal this to you at a later time. If you lose this mnemonic passphrase, you will lose your funds.
                </div>
              </div>
            </div>
            <ul
              class="grid grid-cols-2 gap-x-3 gap-y-5 md:grid-cols-3 lg:grid-cols-4"
            >
              <li
                class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
                data-testid="MnemonicList__item"
              >
                <span
                  class="absolute top-0 left-0 px-1 text-xs font-medium transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
                >
                  1
                </span>
                <span
                  class="font-medium"
                >
                  test
                </span>
              </li>
              <li
                class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
                data-testid="MnemonicList__item"
              >
                <span
                  class="absolute top-0 left-0 px-1 text-xs font-medium transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
                >
                  2
                </span>
                <span
                  class="font-medium"
                >
                  mnemonic
                </span>
              </li>
            </ul>
            <div
              class="Divider-sc-17ehhxk-0 EvPEZ border-theme-secondary-300 dark:border-theme-secondary-800"
              type="horizontal"
            />
            <div
              class="flex justify-between items-center"
            >
              <div
                class="space-y-2"
              >
                <span
                  class="text-lg font-semibold text-theme-secondary-text"
                >
                  Copy or Download Mnemonic Passphrase
                </span>
                <p
                  class="text-sm text-theme-secondary-500"
                >
                  You can copy or download your mnemonic, but store it safely.
                </p>
              </div>
              <div
                class="sc-bdfBwQ gfGMlR"
                height="40"
                width="40"
              >
                <svg>
                  file-password.svg
                </svg>
              </div>
            </div>
            <div
              class="flex justify-end space-x-3"
            >
              <div
                class="relative"
              >
                <button
                  class="ClipboardButton__StyledButton-cammp1-0 kUVLdV"
                  data-testid="SecondSignature__copy"
                  type="button"
                >
                  <div
                    class="flex items-center space-x-2"
                  >
                    <div
                      class="sc-bdfBwQ bcEVDZ"
                      height="16"
                      width="16"
                    >
                      <svg>
                        copy.svg
                      </svg>
                    </div>
                    <span>
                      Copy
                    </span>
                  </div>
                </button>
                <div
                  class="flex absolute inset-0 justify-center items-center rounded bg-theme-primary-100 dark:bg-theme-secondary-800"
                  data-testid="clipboard-button__checkmark"
                  style="opacity: 0;"
                >
                  <div
                    class="sc-bdfBwQ hjDozt text-theme-primary-600 dark:text-theme-secondary-200"
                    height="20"
                    width="20"
                  >
                    <svg>
                      checkmark.svg
                    </svg>
                  </div>
                </div>
              </div>
              <button
                class="OriginalButton__StyledButton-o0fs2j-0 guPjPB"
                data-testid="SecondSignature__download"
                type="button"
              >
                <div
                  class="flex relative items-center space-x-2"
                >
                  <div
                    class="sc-bdfBwQ bcEVDZ"
                    height="16"
                    width="16"
                  >
                    <svg>
                      download.svg
                    </svg>
                  </div>
                  <span>
                    Download
                  </span>
                </div>
              </button>
            </div>
          </div>
          <div
            class="pt-2"
          >
            <div
              class="Divider-sc-17ehhxk-0 EvPEZ border-theme-secondary-300 dark:border-theme-secondary-800"
              type="horizontal"
            />
          </div>
        </section>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SecondSignatureRegistrationForm should render generation step 1`] = `
<DocumentFragment>
  <form
    class="space-y-5"
    data-testid="Form"
  >
    <div>
      <div
        data-testid="tab-pabel__active-panel"
      >
        <section
          data-testid="SecondSignatureRegistrationForm__generation-step"
        >
          <div
            class="flex justify-between items-end bg-theme-background"
          >
            <div
              class="space-y-4"
            >
              <h1
                class="mb-0 text-2xl"
                data-testid="header__title"
              >
                Register Second Signature
              </h1>
              <div
                class="text-theme-secondary-text"
                data-testid="header__subtitle"
              >
                You can additionaly secure your address with a second mnemonic passphrase.
              </div>
            </div>
          </div>
          <div
            class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-warning-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center mt-6"
          >
            <div
              class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-warning-600"
            >
              <div
                class="sc-bdfBwQ hjDozt"
                height="20"
                width="20"
              >
                <svg>
                  alert-warning.svg
                </svg>
              </div>
            </div>
            <div
              class="flex flex-col w-full"
            >
              <div
                class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
              >
                Before creating the second mnemonic, we strongly recommend that you save it, as its loss will lead to a loss of access to your money.
              </div>
            </div>
          </div>
          <div
            class="TransactionDetail__TransactionDetailStyled-r7sfme-0 fyYYPW no-ligatures"
            data-testid="TransactionSender"
          >
            <div
              class="flex-1 space-y-2 w-40 whitespace-nowrap"
            >
              <div
                class="text-sm font-semibold text-theme-secondary-700 no-ligatures"
              >
                Sender
              </div>
              <div
                class="font-semibold"
              >
                <div
                  class="flex overflow-hidden flex-grow items-center space-x-2 whitespace-nowrap no-ligatures"
                >
                  <span
                    class="font-semibold text-base text-theme-text"
                    data-testid="Address__alias"
                  >
                    <span
                      class="TruncateEnd__Wrapper-sc-1hhdnni-0"
                      data-testid="TruncateEnd"
                    >
                      ARK Wallet 1
                    </span>
                  </span>
                  <span
                    class="inline-flex overflow-hidden text-theme-secondary-500 dark:text-theme-secondary-700 font-semibold text-base"
                    data-testid="Address__address"
                  >
                    D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
                  </span>
                </div>
              </div>
            </div>
            <div
              class="flex items-center -space-x-2"
            >
              <div
                class="Avatar__AvatarWrapper-sc-18vw8a4-0 ehnmnK flex-shrink-0"
                data-testid="Avatar"
              >
                <div
                  class="w-full h-full inline-flex items-center justify-center overflow-hidden align-middle rounded-full"
                >
                  <img
                    alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                    src="data:image/svg+xml;utf8,<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" class=\\"picasso\\" width=\\"100\\" height=\\"100\\" viewBox=\\"0 0 100 100\\"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill=\\"rgb(0, 150, 136)\\" width=\\"100\\" height=\\"100\\"/><circle r=\\"50\\" cx=\\"80\\" cy=\\"50\\" fill=\\"rgb(76, 175, 80)\\"/><circle r=\\"35\\" cx=\\"100\\" cy=\\"40\\" fill=\\"rgb(156, 39, 176)\\"/><circle r=\\"40\\" cx=\\"90\\" cy=\\"70\\" fill=\\"rgb(3, 169, 244)\\"/></svg>"
                    title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="pt-6"
          >
            <fieldset
              class="FormField__FormFieldStyled-q4bblx-0 bpBovi flex flex-col space-y-2"
            >
              <label
                class="flex mb-2 text-sm font-semibold transition-colors duration-100 FormLabel text-theme-secondary-text"
                data-testid="FormLabel"
                for="fee"
              >
                Transaction Fee
              </label>
              <div
                class="relative"
                data-testid="InputFee"
              >
                <div
                  class="absolute right-0 -mt-7"
                >
                  <div
                    class="flex items-center"
                  >
                    <span
                      class="Switch__SwitchText-sc-161iiro-0 ddtFSz"
                      role="button"
                    >
                      Simple
                    </span>
                    <label
                      class="flex"
                    >
                      <input
                        class="Toggle__Input-sc-6s3wfu-0 lksfSS"
                        type="checkbox"
                      />
                      <div
                        class="Toggle__Handle-sc-6s3wfu-1 fOBaxM"
                      >
                        <span
                          class="Toggle__HandleInner-sc-6s3wfu-2 jLcWyW"
                        />
                      </div>
                    </label>
                    <span
                      class="Switch__SwitchText-sc-161iiro-0 bxLfsC"
                      role="button"
                    >
                      Advanced
                    </span>
                  </div>
                </div>
                <div
                  class="inline-flex items-center space-x-3 w-full"
                  data-testid="ButtonGroup"
                  role="radiogroup"
                >
                  <button
                    aria-checked="false"
                    aria-label="0"
                    class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ksEKyJ"
                    data-testid="ButtonGroupOption"
                    role="radio"
                    type="button"
                  >
                    <div
                      class="flex flex-col space-y-2"
                    >
                      <div>
                        Slow
                      </div>
                      <span
                        class="text-sm"
                        data-testid="AmountCrypto"
                      >
                        0 DARK
                      </span>
                    </div>
                  </button>
                  <button
                    aria-checked="true"
                    aria-label="1.354"
                    class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ksEKyJ"
                    data-testid="ButtonGroupOption"
                    role="radio"
                    type="button"
                  >
                    <div
                      class="flex flex-col space-y-2"
                    >
                      <div>
                        Average
                      </div>
                      <span
                        class="text-sm"
                        data-testid="AmountCrypto"
                      >
                        1.354 DARK
                      </span>
                    </div>
                  </button>
                  <button
                    aria-checked="false"
                    aria-label="10"
                    class="ButtonGroup__ButtonGroupOptionStyled-sc-2f0qmd-0 ksEKyJ"
                    data-testid="ButtonGroupOption"
                    role="radio"
                    type="button"
                  >
                    <div
                      class="flex flex-col space-y-2"
                    >
                      <div>
                        Fast
                      </div>
                      <span
                        class="text-sm"
                        data-testid="AmountCrypto"
                      >
                        10 DARK
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </fieldset>
          </div>
        </section>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`SecondSignatureRegistrationForm should render transaction details 1`] = `
<DocumentFragment>
  <div
    class="TransactionDetail__TransactionDetailStyled-r7sfme-0 fGJHFM no-ligatures"
    data-testid="TransactionFee"
  >
    <div
      class="flex-1 space-y-2 w-40 whitespace-nowrap"
    >
      <div
        class="text-sm font-semibold text-theme-secondary-700 no-ligatures"
      >
        Transaction Fee
      </div>
      <div
        class="font-semibold"
      >
        <span
          data-testid="AmountCrypto"
        >
          0.1 DARK
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
